### ISSUE TITLE
Visualizing sinc filter kernels

### ISSUE THREAD

**Issue by btlorch** [mravanelli/SincNet#7]
Hi Marco,

thanks for sharing this great work! I'm trying to visualize the weights of the sinc layer in time and frequency domain but I'm having trouble getting it right. Some filters don't look like a single bandpass but have different amplitudes for different frequencies (e.g., see the right example in the eleventh row in the figure below).

The code below just loads a trained model from the saved checkpoint, computes the filters in time domain, and visualizes them alongside their Fourier transform. I'm certain it's just a problem with visualizing them. If you could have a look or share the example from Figure 2 in the paper, that would be highly appreciated.

Thanks,
Benedikt

```python
%matplotlib notebook
import matplotlib.pyplot as plt
import numpy as np
import math
import torch

sampling_frequency = 16000
num_filters = 80
filter_length = 251
frequency_scaling = sampling_frequency * 1.0

# Load trained model from checkpoint
torch_checkpoint = torch.load("exp/SincNet_TIMIT/model_raw.pkl", map_location='cpu')
low_hz = torch_checkpoint["CNN_model_par"]["conv.0.filt_b1"]
band_hz = torch_checkpoint["CNN_model_par"]["conv.0.filt_band"]

# Compute filter kernel. Except from renaming some variables, this should be the same as the original code.
def flip(x, dim):
    xsize = x.size()
    dim = x.dim() + dim if dim < 0 else dim
    x = x.contiguous()
    x = x.view(-1, *xsize[dim:])
    x = x.view(x.size(0), x.size(1), -1)[:,
        getattr(torch.arange(x.size(1) - 1, -1, -1), ('cpu', 'cuda')[x.is_cuda])().long(), :]
    return x.view(xsize)

def sinc(band, t_right):
    y_right = torch.sin(2 * math.pi * band * t_right) / (2 * math.pi * band * t_right)
    y_left = flip(y_right, 0)

    y = torch.cat([y_left, torch.autograd.Variable(torch.ones(1)), y_right])

    return y

filters = torch.autograd.Variable(torch.zeros((num_filters, filter_length)))
N = filter_length
t_right = torch.autograd.Variable(torch.linspace(1, (N - 1) / 2, steps=int((N - 1) / 2)) / sampling_frequency)

min_freq = 50.0;
min_band = 50.0;

filt_beg_freq = torch.abs(low_hz) + min_freq / frequency_scaling
filt_end_freq = filt_beg_freq + (torch.abs(band_hz) + min_band / frequency_scaling)

n = torch.linspace(0, N, steps=N)

# Filter window (hamming)
window = 0.54 - 0.46 * torch.cos(2 * math.pi * n / N);
window = torch.autograd.Variable(window.float())

for i in range(num_filters):
    low_pass1 = 2 * filt_beg_freq[i].float() * sinc(filt_beg_freq[i].float() * frequency_scaling, t_right)
    low_pass2 = 2 * filt_end_freq[i].float() * sinc(filt_end_freq[i].float() * frequency_scaling, t_right)
    band_pass = (low_pass2 - low_pass1)

    band_pass = band_pass / torch.max(band_pass)

    filters[i, :] = band_pass * window

filters = filters.view(num_filters, 1, filter_length)
filters = filters.detach().numpy()

# Visualize filter kernels (similar to https://gist.github.com/endolith/236567)
# Two filters and their Fourier transform per row.
num_cols = 4
num_rows = int(np.ceil(num_filters * 2 / num_cols))

fig, axes = plt.subplots(num_rows, num_cols, figsize=(9, 80))
for i in range(num_filters):
    spatial_ax = axes[(i * 2) // 4, (i * 2) % 4]
    frequency_ax = axes[(i * 2 + 1) // 4, (i * 2 + 1) % 4]
    
    weights = filters[i, 0, :]
    
    # Frequency computation
    ampl = 1/N * np.abs(np.fft.rfft(weights))
    
    # RFFT frequency bins
    freqs = np.fft.rfftfreq(N, 1/sampling_frequency)
    
    spatial_ax.plot(weights)
    frequency_ax.stem(freqs, ampl)
    
fig.tight_layout()
```

![visualize_weights](https://user-images.githubusercontent.com/11959715/46261723-0636fa80-c4f8-11e8-87ef-0c55b2102e92.png)

**Comment by mravanelli** [comment]
Dear Benedikt,
please forgive my delay in answering.  I haven't  observed such weird filter shapes in my TIMIT experiments. 
A possible explanation  is that for same filters the end_frequency ends up to be higher than the Nyquist frequency.  Take a look into the following lines of code:
```python
filt_beg_freq=torch.abs(self.filt_b1)+min_freq/self.freq_scale
filt_end_freq=filt_beg_freq+(torch.abs(self.filt_band)+min_band/self.freq_scale)
```
The filt_end_freq can be in principle any possible value, even higher that 0.5 (that is the normalized Nyquist frequnecy). If that happens the filter can take some kind of weird shape. If the current release code I didn't add any threshold for that since in my case I haven't seen filters exceeding the Nyquist freq. If this happens rarely, I don't think it might hurt the performance. It could be anyway good to add a clip into the code.

**Comment by btlorch** [comment]
Thanks Mirco, I'll give this a try should I have to retrain the network at some point. It looks like it doesn't hurt the performance without clamping the filters, though.
